Improve some use of Regex

stephentoub opened the pull request:
- Use IsMatch instead of Match(...).Success
- Use EnumerateMatches when Index/Length are needed but not captures
- Remove Regex that could just be IndexOf
- Regex.Match never returns null, so remove null checks

joperezr commented on src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/RegularExpressionAttribute.cs:
```diff
@@ -66,11 +66,14 @@ public override bool IsValid(object? value)
             }
 
-            var m = Regex!.Match(stringValue);
+            foreach (ValueMatch m in Regex!.EnumerateMatches(stringValue))
```
NIT: since we are only using the first result of the enumerator, would it make sense to instead call MoveNext/Current directly?

stephentoub replied:
I've gone back and forth on this one. The generated code is slightly smaller without a loop, but it didn't have a measurable impact and I personally find the foreach easier to read and reason about, so I'd like to leave it.

joperezr reviewed (approved):
Left a small comment but looks good otherwise. I wonder if we should add another analyzer that suggests `IsMatch` for people that are only doing `Match().Success`

joperezr commented:
Lol just saw you logged https://github.com/dotnet/runtime/issues/68962 this morning, so that covers my comment about probably wanting an analyzer 😆

stephentoub committed:
Improve some use of Regex

- Use IsMatch instead of Match(...).Success
- Use EnumerateMatches when Index/Length are needed but not captures
- Remove Regex that could just be IndexOf
